### PR TITLE
Restructure PerformanceCalculator to not require ScoreInfo argument

### PR DESCRIPTION
**Pull request by smoogipoo**
In order to reduce the number of performance calculators created (such as on [every judgement via `PerformancePointsCounter`](https://github.com/ppy/osu/blob/ffd87e78af2932b00866b907873c6ef777ed3f38/osu.Game/Screens/Play/HUD/PerformancePointsCounter.cs#L117-L137)).

I've also removed the `TimeRate` value, as it's not used by any ruleset for the time being. There's more cleanups that can be done to in order to promote thread-safety as future work if required - by making the classes completely stateless, but I see that as unnecessary work for the time being.

In terms of structure ("why `CreatePerformanceAttributes`?"), it's because I'm following the difficulty calculator API.

Contains breaking changes.

# vNext

## `PerformanceCalculator` method signatures have changed

The `PerformanceCalculator` signature has changed a bit in order for performance calculators to be constructed less during critical code paths. Refer to the following diff for required changes:
```diff
// Constructor signature.
- ctor(Ruleset ruleset, DifficultyAttributes attributes, ScoreInfo score);
+ ctor();

// These are the methods that actually perform the calculation:
- public virtual PerformanceAttributes Calculate();
+ protected virtual PerformanceAttributes CreatePerformanceAttributes(ScoreInfo score, DifficultyAttributes attributes);

// Helper methods to calculate performance:
+ public PerformanceAttributes Calculate(ScoreInfo score, DifficultyAttributes attributes);
+ public PerformanceAttributes Calculate(ScoreInfo score, IWorkingBeatmap beatmap);

// Ruleset API:
- public PerformanceCalculator CreatePerformanceCalculator(IWorkingBeatmap beatmap, ScoreInfo score);
- public virtual PerformanceCalculator CreatePerformanceCalculator(DifficultyAttributes attributes, ScoreInfo score);
+ public virtual PerformanceCalculator CreatePerformanceCalculator();
```

### CODE REVIEW

**Review by frenzibyte — DISMISSED** [review]
Looks correct on a line-by-line check.

**Review by bdach — DISMISSED** [review]
haven't scrutinised this too hard but looks correct on a moderately detailed readthrough

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs [hunk 1]
```diff
@@ -14,37 +14,35 @@ namespace osu.Game.Rulesets.Taiko.Difficulty
...
-        private Mod[] mods;
         private int countGreat;
         private int countOk;
         private int countMeh;
         private int countMiss;
 
-        public TaikoPerformanceCalculator(Ruleset ruleset, DifficultyAttributes attributes, ScoreInfo score)
-            : base(ruleset, attributes, score)
+        public TaikoPerformanceCalculator(Ruleset ruleset)
+            : base(ruleset)
```
Comment (peppy) [inline osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs:23]: I'm wondering if this needs the ruleset input. ie. is there any reason to not do:

```suggestion
        public TaikoPerformanceCalculator() : base(new TaikoRuleset())
```

Comment (smoogipoo) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs:23]: Sure

Comment (smoogipoo) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs]: Also removed the similar argument in the ScoreProcessor PR in https://github.com/ppy/osu/pull/17243/commits/daac93349864ee975aab450fd6b7d4b48e77bac6

Comment (peppy) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs]: I'm not 100% on this change but I guess it's probably fine as long as `Ruleset` doesn't contain any dynamic pieces going forward (and as long as we don't mind one extra `RulesetInfo` being constructed). Let's see how things pan out.

**Review by peppy — APPROVED** [review]

**Pull request merged**